pandas/core/internals/concat.py

```diff
@@ -93,6 +102,80 @@ def _concatenate_array_managers(
         return ArrayManager(arrays, [axes[1], axes[0]], verify_integrity=False)
 
 
+def concat_arrays(to_concat: List[Any]) -> ArrayLike:
+    """
+    Alternative for concat_compat but specialized for use in the ArrayManager.
+
+    Differences: only deals with 1D arrays (no axis keyword), assumes
+    ensure_wrapped_if_datetimelike and does not skip empty arrays to determine
+    the dtype.
+    In addition ensures that all NullArrayProxies get replaced with actual
+    arrays.
+
+    Parameters
+    ----------
+    to_concat : list of arrays
+
+    Returns
+    -------
+    np.ndarray or ExtensionArray
+    """
+    # ignore the all-NA proxies to determine the resulting dtype
+    to_concat_no_proxy = [x for x in to_concat if not isinstance(x, NullArrayProxy)]
+
+    kinds = {obj.dtype.kind for obj in to_concat_no_proxy}
+    single_dtype = len({x.dtype for x in to_concat_no_proxy}) == 1
+    any_ea = any(is_extension_array_dtype(x.dtype) for x in to_concat_no_proxy)
+
+    if any_ea:
+        if not single_dtype:
+            target_dtype = find_common_type([x.dtype for x in to_concat_no_proxy])
+        else:
+            target_dtype = to_concat_no_proxy[0].dtype
+    elif not single_dtype:
+        if any(kind in ["m", "M"] for kind in kinds):
```

Comment:
why cant we use find_common_type here?

Reply:
Good question. So, can we think of any combination of dtypes where `np.find_common_type(..)` (which would get used by our find_common_type in this case) would not result in object dtype if the input contains datetime or timedelta?

Most obvious case I could think of is with integers, and numpy also gives object dtype there:

```
In [20]: np.find_common_type([np.dtype("datetime64[ns]"), np.dtype("int64")], [])
Out[20]: dtype('O')
```

The reason I handled it separately here is because in `dtypes/concat.py` we also handle datetime-likes separately, and in `_concat_datetime` we also "hardcode" to use object dtype in case of multiple input dtypes. 
But if `find_common_dtype` always gives object dtype for mixed datetimelike input, this can indeed be simplified here. Will try that and see what the tests say.

Reply:
@jbrockmendel so I was indeed able to simplify this, also the `if any_ea` check was not needed, since `find_common_type`  handles all cases